util/cuda/include/cuda_ptr.h

```diff
@@ -32,8 +32,27 @@ class CudaPtr {
     }
     CudaPtr() = default;
     ~CudaPtr() { free(); }
-    CudaPtr(const CudaPtr&) = delete;  // class does not support copying(and moving)
+    CudaPtr(const CudaPtr&) = delete;  // class does not support copying
     CudaPtr& operator=(const CudaPtr&) = delete;
+    CudaPtr(CudaPtr<T>&& oth) noexcept
+        : device_ptr_(oth.device_ptr_), elem_count_(oth.elem_count_), capacity_(oth.capacity_) {
+        oth.device_ptr_ = nullptr;
+        oth.elem_count_ = oth.capacity_ = 0;
+    }
+
+    CudaPtr& operator=(CudaPtr<T>&& rhs) noexcept {
+        if (&rhs != this) {
+            if (device_ptr_) {
+                free();
```

Comment:
Add "REPORT_WHEN_CUDA_ERR()" to "cudaFree()" call inside the free().

Reply:
Or if one wants to be fancy, make free() return cuda status. So one can make it throw in the assignment operator, but warn inside the dtor.